src/vs/vscode.proposed.d.ts

```diff
@@ -1660,10 +1653,22 @@ declare module 'vscode' {
 	 * modified with the methods on the run task.
 	 */
 	export interface NotebookCellExecutionTask {
+		readonly document: NotebookDocument;
+		readonly cell: NotebookCell;
+
 		start(context?: NotebookCellExecuteStartContext): void;
+		setExecutionOrder(order: number): void;
 		end(result: NotebookCellPreviousExecutionResult): void;
-		token: CancellationToken;
+		readonly token: CancellationToken;
 
+		getOutputEdit(): NotebookCellOutputEdit;
+		applyOutputEdit(edit: NotebookCellOutputEdit): Thenable<boolean>;
+	}
+
+	/**
+	 * TODO@rob Similar to WorkspaceEdit, does it need other methods from that?
+	 */
+	export interface NotebookCellOutputEdit {
```

Comment:
I don't like this tbh. All output changes on the execution task should be live, no transaction, no nothing but go directly on the renderer model

Reply:
This came from discussion with Don. They need to know when an output has been applied because of some workarounds they have that involve outputs that depend on previous outputs (terminal control codes in streaming outputs, which we don't support) which means they need the same sort of async apply API that workspace edits have. They need to await a write then read with the write applied, I think it makes sense to be able to do that.

Reply:
>  outputs that depend on previous outputs (terminal control codes in streaming outputs, which we don't support)

Yeah, just talked to him and we covered this. Basically there can be stdout/err output with ANSI sequences. We support some of those sequences, like colours, but we don't support the one that says "clear screen, start at the top". I strongly feel that we should support those sequences. It will make their live much simpler as they don't need to scan outputs anymore and for us it should be simple to implement (find sequence, clear html node). 

> they need to know when an output has been applied because

I am generally not against letting extension know but a slicker API would be to let those append/replace/clear-functions return a promise which signals when/if the change happened. Tho, it not happening would be a bad bug on our end

Reply:
created https://github.com/microsoft/vscode/issues/118833 for the ANSI sequence gaps

Reply:
> would be to let those append/replace/clear-functions return a promise which signals when/if the change happened

I don't have a strong opinion about doing this vs something that looks more like a workspace edit. It basically is like a workspace edit but maybe the user wouldn't see it that way?

Reply:
>  It basically is like a workspace edit but maybe the user wouldn't see it that way?

I'd say it one level below a workspace edit, it doesn't go onto the undo/redo stack and won't be handle by the preview workspace edit feature (which today is broken). It's merely a model editor of the notebook model